File: ts/step5_tco.js
"use strict";
var printer_1 = require("./printer");
var reader_1 = require("./reader");
var env_1 = require("./env");
var core_1 = require("./core");
var readline = require("readline");
var rl = readline.createInterface({
    input: process.stdin,
    output: process.stdout,
    terminal: true
});
var repl_env = new env_1.Env();
repl_env.set("+", {
    type: "function",
    value: {
        fn: function (a, b) { return ({
            type: "number",
            value: a.value + b.value
        }); }
    }
});
repl_env.set("-", {
    type: "function",
    value: {
        fn: function (a, b) { return ({
            type: "number",
            value: a.value - b.value
        }); }
    }
});
repl_env.set("*", {
    type: "function",
    value: {
        fn: function (a, b) { return ({
            type: "number",
            value: a.value * b.value
        }); }
    }
});
repl_env.set("/", {
    type: "function",
    value: {
        fn: function (a, b) { return ({
            type: "number",
            value: a.value / b.value
        }); }
    }
});
Object.entries(core_1.ns).forEach(function (_a) {
    var key = _a[0], value = _a[1];
    return repl_env.set(key, value);
});
var eval_ast = function (ast, env) {
    if (ast.type === "symbol") {
        try {
            var val = env.get(ast.value);
            return val;
        }
        catch (err) {
            return {
                type: "error",
                value: ".*'" + ast.value + "' not found.*'"
            };
        }
    }
    if (ast.type === "list") {
        return {
            type: "list",
            value: ast.value.map(function (val) { return EVAL(val, env); })
        };
    }
    return ast;
};
var EVAL = function (ast, env) {
    var _loop_1 = function () {
        var _a;
        if (ast.type !== "list") {
            return { value: eval_ast(ast, env) };
        }
        if (ast.value.length === 0) {
            return { value: ast };
        }
        if (ast.value[0].value === "def!") {
            if (ast.value[1].type !== "symbol") {
                throw new Error("Should be a symbol");
            }
            var evaluated_1 = EVAL(ast.value[2], env);
            if (evaluated_1.type !== "error") {
                env.set(ast.value[1].value, evaluated_1);
            }
            return { value: evaluated_1 };
        }
        if (ast.value[0].value === "let*") {
            var new_env = new env_1.Env(env);
            if (ast.value[1].type !== "list") {
                throw new Error("Bindings should be a list");
            }
            var bindings = ast.value[1].value;
            if (bindings.length % 2 !== 0) {
                throw new Error("Bindings should not be odd length");
            }
            var i = 0;
            while (i < bindings.length) {
                if (bindings[i].type !== "symbol") {
                    throw new Error("Bindings should be a symbol");
                }
                else {
                    new_env.set(bindings[i].value, EVAL(bindings[i + 1], new_env));
                }
                i += 2;
            }
            ast = ast.value[2];
            env = new_env;
            return "continue";
        }
        if (ast.value[0].value === "do") {
            var evaluated_2 = ast.value.slice(1, -1).map(function (val) { return EVAL(val, env); });
            ast = ast.value[ast.value.length - 1];
            return "continue";
        }
        if (ast.value[0].value === "if") {
            var cond = EVAL(ast.value[1], env);
            if (cond.value !== false && cond.type !== "nil") {
                ast = ast.value[2];
            }
            else if (ast.value.length === 4) {
                ast = ast.value[3];
            }
            else {
                ast = {
                    type: "nil"
                };
            }
            return "continue";
        }
        if (ast.value[0].value === "fn*") {
            var bindings_1 = ast.value[1];
            var content_1 = ast.value[2];
            if (bindings_1.type !== "list" || content_1.type !== "list") {
                return { value: {
                        type: "error",
                        value: "Function bindings and content should be a list"
                    } };
            }
            return { value: {
                    type: "function",
                    value: {
                        params: bindings_1,
                        ast: content_1,
                        env: env,
                        fn: function () {
                            var args = [];
                            for (var _i = 0; _i < arguments.length; _i++) {
                                args[_i] = arguments[_i];
                            }
                            try {
                                var new_env = new env_1.Env(env);
                                for (var i = 0; i < args.length; i++) {
                                    if (bindings_1.value[i].type !== "symbol") {
                                        return {
                                            type: "error",
                                            value: "Function bindings should all be args"
                                        };
                                    }
                                    new_env.set(bindings_1.value[i].value, args[i]);
                                }
                                return EVAL(content_1, new_env);
                            }
                            catch (e) {
                                return {
                                    type: "error",
                                    value: "Unexpected error"
                                };
                            }
                        }
                    }
                } };
        }
        var evaluated = eval_ast(ast, env);
        if (evaluated.type !== "list") {
            return { value: {
                    type: "error",
                    value: "Should be a function"
                } };
        }
        if (evaluated.value[0].type === "error") {
            return { value: evaluated.value[0] };
        }
        if (evaluated.value[0].type !== "function") {
            return { value: {
                    type: "error",
                    value: "Should be a function"
                } };
        }
        var fnValue = evaluated.value[0].value;
        if (fnValue.params) {
            var args = evaluated.value.slice(1);
            ast = fnValue.ast;
            var new_env = new env_1.Env(env);
            for (var i = 0; i < args.length; i++) {
                if (fnValue.params.value[i].type !== "symbol") {
                    return { value: {
                            type: "error",
                            value: "Function bindings should all be args"
                        } };
                }
                new_env.set(fnValue.params.value[i].value, args[i]);
            }
            env = new_env;
            return "continue";
        }
        return { value: (_a = evaluated.value[0].value).fn.apply(_a, evaluated.value.slice(1)) };
    };
    while (true) {
        var state_1 = _loop_1();
        if (typeof state_1 === "object")
            return state_1.value;
    }
};
var rep = function (input) { return printer_1.pr_str(EVAL(reader_1.read_str(input), repl_env)); };
process.stdout.write("user> ");
rl.on("line", function (line) {
    console.log(rep(line));
    process.stdout.write("user> ");
});
module.exports = {};

File: impls/python/step5_tco.py
import sys
from reader import read_str, Val
from printer import pr_str
from env import Env
from core import ns


repl_env = Env()
for key in ns:
    repl_env.set(key, ns[key])

def eval_ast(ast, env):
    if ast.type == "symbol":
        return env.get(ast.value)
    if ast.type in ("list", "vector"):
        items = map(lambda ast: eval(ast, env), ast.value)
        return Val(ast.type, items)
    if ast.type == "hashmap":
        new_val = {}
        for key, value in ast.value.items():
            new_val[key] = eval(value, env)
        return Val("hashmap", new_val)
    return ast

def eval(ast, env):
    while True:
        if ast.type != "list":
            return eval_ast(ast, env)
        if len(ast.value) == 0:
            return ast

        symbol = ast.value[0].value
        if symbol == "def!":
            evaluated = eval(ast.value[2], env)
            env.set(ast.value[1].value, evaluated)
            return evaluated

        if symbol == "let*":
            new_env = Env(env)
            params = ast.value[1].value

            i = 0
            while i < len(params):
                new_env.set(params[i].value, eval(params[i + 1], new_env))
                i += 2
            ast = ast.value[2]
            env = new_env
            continue

        if symbol == "do":
            evaluated = None
            for i in ast.value[1:-1]:
                eval(i, env)
            ast = ast.value[len(ast.value) -1]
            continue

        if symbol == "if":
            evaluated = None
            cond = eval(ast.value[1], env)
            if cond.type != "nil" and (cond.type != "bool" or cond.value == True):
                ast = ast.value[2]
            elif len(ast.value) > 3:
                ast = ast.value[3]
            else: 
                ast = Val("nil", [])
            continue

        if symbol == "fn*":
            def func(*exprs_t):
                new_env = Env(env)
                params = ast.value[1].value
                exprs = list(exprs_t)
                i = 0
                while i < len(params):
                    if params[i].value == "&":
                        new_env.set(params[i + 1].value, Val("list", exprs[i:]))
                        break
                    new_env.set(params[i].value, exprs[i])
                    i += 1

                return eval(ast.value[2], new_env)

            return Val("custom_fn", {
                "fn": func,
                "ast": ast.value[2],
                "params": ast.value[1].value,
                "env": env
            })

        new_ast = eval_ast(ast, env)

        fn = new_ast.value[0]
        if fn.type == 'custom_fn':
            exprs = new_ast.value[1:]
            ast = fn.value["ast"]
            params = fn.value["params"]
            i = 0
            new_env = Env(fn.value["env"])
            while i < len(params):
                if params[i].value == "&":
                    new_env.set(params[i + 1].value, Val("list", exprs[i:]))
                    break
                new_env.set(params[i].value, exprs[i])
                i += 1
            env = new_env
            continue

        return fn.value(*new_ast.value[1:])

def rep(line):
    try:
        print(pr_str(eval(read_str(line), repl_env)))
    except Exception as e:
        if isinstance(e.args[0], Val):
            print(pr_str(Val("error", e.args[0])))
        else:
            raise e
            print(e.traceback)
            print(pr_str(Val("error", Val("string", str(e)))))

rep("(def! not (fn* (a) (if a false true)))")
while True:
    line = raw_input("user> ")
    if (len(line) != 0):
        rep(line)
